src/runtime_src/core/common/usage_metrics.h

```diff
@@ -0,0 +1,167 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2023 Advanced Micro Devices, Inc. All rights reserved.
+#ifndef XRT_CORE_USAGE_METRICS_H
+#define XRT_CORE_USAGE_METRICS_H
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <stdint.h>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <vector>
+
+#include "core/include/ert.h"
+#include "core/include/xrt.h"
+#include "core/include/xrt/xrt_hw_context.h"
+#include "core/include/xrt/xrt_uuid.h"
+
+// forward declaration of xrt_core::device class
+namespace xrt_core {
+class device;
+}
+
+namespace {
+using tp = std::chrono::time_point<std::chrono::high_resolution_clock>;
+}
+
+////////////////////////////////////////////////////////////////
+// namespace xrt_core::usage_metrics
+//
+// Used for printing XRT usage metrics of Application.
+//
+// This information is printed at the end of the application run
+// and the infrastructure must be initialized before launching the
+// application using xrt.ini
+//
+// % cat xrt.ini
+// [Runtime]
+// usage_metrics_logging = true
+////////////////////////////////////////////////////////////////
+namespace xrt_core::usage_metrics {
+
+struct bo_metrics
+{
+  uint32_t total_count = 0;
+  size_t   total_size_in_bytes = 0;
+  size_t   peak_size_in_bytes = 0;
+  size_t   bytes_synced_to_device = 0;
+  size_t   bytes_synced_from_device = 0;
+};
+
+struct kernel_metrics
+{
+  struct timestamp
+  {
+    tp start_time;
+    bool is_valid = false;
+  };
+
+  std::string name;
+  std::vector<uint32_t> cu_index_vec;
+  uint32_t total_runs = 0;
+  std::chrono::microseconds total_time = {};
+  std::unordered_map<uintptr_t, timestamp> exec_times; // run handle ptr is used for indexing
+  size_t num_args;
+};
+
+struct hw_ctx_metrics
+{
+  uintptr_t id;  // using hw_ctx handle ptr as unique identifier for logging
+  xrt::uuid xclbin_uuid;
+  struct bo_metrics bos_met;
+  std::vector<struct kernel_metrics> kernel_metrics_vec;
+};
+
+struct device_metrics
+{
+  std::string bdf = "";
+  struct bo_metrics global_bos_met;
+  uint32_t bo_active_count = 0;
+  uint32_t bo_peak_count = 0;
+  std::vector<struct hw_ctx_metrics> hw_ctx_vec;
+};
+
+using metrics_map = std::map<std::thread::id, std::map<unsigned int, device_metrics>>;
+
+// class base_logger - class with no op calls
+//
+// when user doesn't set ini option logging should be no op
+class base_logger
+{
+public:
+  virtual void 
+  log_device_info(std::shared_ptr<xrt_core::device>) {}
+
+  virtual void
+  log_hw_ctx_info(void*) {}
+
+  virtual void 
+  log_buffer_info_construct(unsigned int, size_t, void*) {}
+  
+  virtual void 
+  log_buffer_info_destruct(unsigned int) {}
+
+  virtual void
+  log_buffer_sync(unsigned int, void*, size_t, xclBOSyncDirection) {}
+
+  virtual void
+  log_kernel_info(std::shared_ptr<xrt_core::device>, const xrt::hw_context&, const std::string&, size_t) {}
+
+  virtual void
+  log_kernel_run_info(void*, const void*, bool, ert_cmd_state) {}
```

Comment:
I am not sure it worthwhile to go to void*, pass actual objects by const reference is probably better.